src/Build.UnitTests/Graph/IsolateProjects_Tests.cs

```diff
@@ -59,51 +90,248 @@ public void CacheEnforcementShouldAcceptSelfReferences(BuildResultCode expectedB
 
                     logger.ErrorCount.ShouldBe(1);
 
-                    logger.Errors.First().Message.ShouldStartWith("MSB4252:");
-                });
+                    logger.Errors.First()
+                        .Message.ShouldStartWith("MSB4252:");
+                },
+                addContinueOnError: addContinueOnError);
         }
 
         [Fact]
         public void CacheEnforcementShouldAcceptPreviouslyBuiltReferences()
         {
-            AssertBuild(new []{"BuildReference"},
+            AssertBuild(new []{"BuildDeclaredReference"},
                 (result, logger) =>
                 {
                     result.OverallResult.ShouldBe(BuildResultCode.Success);
 
                     logger.Errors.ShouldBeEmpty();
                 },
-                buildReference: true);
+                buildDeclaredReference: true);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void TaskEnforcementShouldFailOnUndeclaredReference(bool addContinueOnError)
+        {
+            AssertBuild(new[] { "BuildUndeclaredReference" },
+                (result, logger) =>
+                {
+                    result.OverallResult.ShouldBe(BuildResultCode.Failure);
+
+                    logger.ErrorCount.ShouldBe(1);
+
+                    logger.Errors.First().Message.ShouldStartWith("MSB4254:");
+                },
+                addContinueOnError: addContinueOnError);
         }
 
-        private void AssertBuild(string[] targets, Action<BuildResult, MockLogger> assert, bool buildReference = false)
+        [Fact]
+        public void TaskEnforcementShouldNotAcceptPreviouslyBuiltReferences()
+        {
+            AssertBuild(new[] { "BuildUndeclaredReference" },
+                (result, logger) =>
+                {
+                    result.OverallResult.ShouldBe(BuildResultCode.Failure);
+
+                    logger.ErrorCount.ShouldBe(1);
+
+                    logger.Errors.First().Message.ShouldStartWith("MSB4254:");
+                },
+                buildUndeclaredReference: true);
+        }
+
+        public static IEnumerable<object[]> TaskEnforcementShouldNormalizeFilePathsTestData
+        {
+            get
+            {
+                Func<string, string> preserve = path => path;
+
+                Func<string, string> fullToRelative = path =>
+                {
+                    var directory = Path.GetDirectoryName(path);
+                    var file = Path.GetFileName(path);
+
+                    return Path.Combine("..", directory, file);
+                };
+
+                Func<string, string> ToForwardSlash = path => path.ToSlash();
+
+                Func<string, string> ToBackwardSlash = path => path.ToBackwardSlash();
+
+                Func<string, string> ToDuplicateSlashes = path => path.Replace("/", "//").Replace(@"\", @"\\");
+
+                yield return new object[]
```

Comment:
If this is the full matrix, can you compute it dynamically instead for clarity? If it's not, can you comment what it is and why it's not?